{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SCENIC analysis - Template script\n",
    "\n",
    "*Last updated*: 22 jan 2019\n",
    "\n",
    "(For a detailed tutorial with explanations of each step see: https://pyscenic.readthedocs.io )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before starting: Data loading/formatting and gene filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set dataset location & organism. In this example we are using:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Dataset*: __Mouse brain__ (somatosensory cortex and hippocampus); 3005 cells.\n",
    "\n",
    "```\n",
    "wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE60nnn/GSE60361/suppl/GSE60361%5FC1%2D3005%2DExpression%2Etxt%2Egz ; gzip -d GSE60361_C1-3005-Expression.txt.gz\n",
    "```\n",
    "\n",
    "> A. Zeisel, A. B. Moz-Manchado, S. Codeluppi, P. Lönnerberg, G. L. Manno, A. Juréus, S. Marques, H. Munguba, L. He, C. Betsholtz, C. Rolny, G. Castelo-Branco, J. Hjerling-Leffler, and S. Linnarsson, “Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq,” Science, vol. 347, no. 6226, pp. 1138–1142, Mar. 2015.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "exprMat_path = 'GSE60361_C1-3005-Expression.txt'\n",
    "exprMat_filtered_path = 'filteredExpressionMatrix.loom' # output file, choose name...\n",
    "\n",
    "ORGANISM = 'mouse' # Select from: 'human' / 'mouse'/ 'drosophila'\n",
    "DATABASE_FOLDER = \"databases/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A motif database for the relevant organism will be used to check for existing genes.\n",
    "\n",
    "Edit for using a different version (all databases of same version/organism contain the same genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "databases/mm9-500bp-upstream-7species.mc9nr.feather\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "if(ORGANISM == 'human'):\n",
    "    DATABASES_GLOB = os.path.join(DATABASE_FOLDER, \"hg19-500bp-upstream-7species.mc9nr.feather\")\n",
    "if(ORGANISM == 'mouse'):\n",
    "    DATABASES_GLOB = os.path.join(DATABASE_FOLDER, \"mm9-500bp-upstream-7species.mc9nr.feather\")\n",
    "if(ORGANISM == 'drosophila'):\n",
    "    DATABASES_GLOB = os.path.join(DATABASE_FOLDER, \"dm6-5kb-upstream-full-tx-11species.mc8nr.feather\")\n",
    "    \n",
    "print(DATABASES_GLOB)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The expression matrix of this example is stored in a text file (as tab separated matrix).\n",
    "\n",
    "To load 10X expression files (e.g. CellRanger output in `.h5`), scripts are available at their website (https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/python). For example, using the matrix `get_matrix_from_h5` from http://cf.10xgenomics.com/supp/cell-exp/megacell_tutorial-1.0.1.htm`: \n",
    "```\n",
    "tmp = get_marix_from_h5(exprMat_path, 'dm6')\n",
    "exprMat_full=pd.DataFrame(data=tmp.matrix.toarray().squeeze(),\n",
    "                 index=tmp.gene_names.astype('str'),\n",
    "                 columns=tmp.barcodes.astype('str')) \n",
    "del(tmp)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyarrow.feather as feather"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From text file: \n",
    "exprMat = pd.read_csv(exprMat_path, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "exprMat = exprMat.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19972, 3005)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1772071015_C02</th>\n",
       "      <th>1772071017_G12</th>\n",
       "      <th>1772071017_A05</th>\n",
       "      <th>1772071014_B06</th>\n",
       "      <th>1772067065_H06</th>\n",
       "      <th>1772071017_E02</th>\n",
       "      <th>1772067065_B07</th>\n",
       "      <th>1772067060_B09</th>\n",
       "      <th>1772071014_E04</th>\n",
       "      <th>1772071015_D04</th>\n",
       "      <th>...</th>\n",
       "      <th>1772066110_D12</th>\n",
       "      <th>1772071017_A07</th>\n",
       "      <th>1772063071_G10</th>\n",
       "      <th>1772058148_C03</th>\n",
       "      <th>1772063061_D09</th>\n",
       "      <th>1772067059_B04</th>\n",
       "      <th>1772066097_D04</th>\n",
       "      <th>1772063068_D01</th>\n",
       "      <th>1772066098_A12</th>\n",
       "      <th>1772058148_F03</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tspan12</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tshz1</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fnbp1l</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Adamts15</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cldn12</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3005 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          1772071015_C02  1772071017_G12  1772071017_A05  1772071014_B06  \\\n",
       "cell_id                                                                    \n",
       "Tspan12                0               0               0               3   \n",
       "Tshz1                  3               1               0               2   \n",
       "Fnbp1l                 3               1               6               4   \n",
       "Adamts15               0               0               0               0   \n",
       "Cldn12                 1               1               1               0   \n",
       "\n",
       "          1772067065_H06  1772071017_E02  1772067065_B07  1772067060_B09  \\\n",
       "cell_id                                                                    \n",
       "Tspan12                0               0               3               0   \n",
       "Tshz1                  2               2               2               1   \n",
       "Fnbp1l                 1               2               1               0   \n",
       "Adamts15               0               0               0               0   \n",
       "Cldn12                 0               0               0               0   \n",
       "\n",
       "          1772071014_E04  1772071015_D04       ...        1772066110_D12  \\\n",
       "cell_id                                        ...                         \n",
       "Tspan12                0               0       ...                     0   \n",
       "Tshz1                  0               2       ...                     0   \n",
       "Fnbp1l                 5               2       ...                     0   \n",
       "Adamts15               0               0       ...                     0   \n",
       "Cldn12                 2               3       ...                     0   \n",
       "\n",
       "          1772071017_A07  1772063071_G10  1772058148_C03  1772063061_D09  \\\n",
       "cell_id                                                                    \n",
       "Tspan12                0               0               0               0   \n",
       "Tshz1                  0               0               0               0   \n",
       "Fnbp1l                 0               0               0               0   \n",
       "Adamts15               0               0               0               0   \n",
       "Cldn12                 0               0               0               0   \n",
       "\n",
       "          1772067059_B04  1772066097_D04  1772063068_D01  1772066098_A12  \\\n",
       "cell_id                                                                    \n",
       "Tspan12                0               0               0               0   \n",
       "Tshz1                  0               0               0               0   \n",
       "Fnbp1l                 0               0               0               0   \n",
       "Adamts15               0               0               0               0   \n",
       "Cldn12                 0               0               0               0   \n",
       "\n",
       "          1772058148_F03  \n",
       "cell_id                   \n",
       "Tspan12                1  \n",
       "Tshz1                  1  \n",
       "Fnbp1l                 0  \n",
       "Adamts15               0  \n",
       "Cldn12                 0  \n",
       "\n",
       "[5 rows x 3005 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exprMat.shape\n",
    "exprMat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene filter\n",
    "Check whether these filters are appropriate for your dataset.\n",
    "Adjust the minimum counts and number of cells if needed (e.g. `minCountsPerGene` and `minSamples`).\n",
    "\n",
    "These are the same filters as the original [SCENIC implementation in R](https://github.com/aertslab/SCENIC) (expained in the tutorial [\"Running SCENIC\"](https://rawcdn.githack.com/aertslab/SCENIC/master/inst/doc/SCENIC_Running.html))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General stats: \n",
      "\t nGenes detected (counts>0): 19972\n",
      "\t nGenesPerCell:  770  -  8146\n",
      "\t Maximum value in matrix:  10738\n",
      "\t Number of counts (in the dataset units) per gene: 2  -  2200118\n",
      "\t Number of cells in which each gene is detected: 1  -  3004\n",
      "Used for filtering: \n",
      "\t minCountsPerGene:  27.05\n",
      "\t minSamples:  9.02  (nCells: 3005 )\n",
      "Genes that pass the filter:\n",
      "\t  15074 \tgenes with counts per gene >  27.05\n",
      "\t  15072 \tgenes detected in more than  9.02  cells\n",
      "\t  13800 \tgenes are in the motif database (the database contains info for  22058 genes)\n"
     ]
    }
   ],
   "source": [
    "## General stats\n",
    "nGenesPerCell = np.sum(exprMat>0, axis=0)\n",
    "nCountsPerGene = np.sum(exprMat, axis=1)\n",
    "nCellsPerGene = np.sum(exprMat>0, axis=1)\n",
    "nCells=exprMat.shape[1]\n",
    "\n",
    "print(\"General stats: \")\n",
    "print(\"\\t nGenes detected (counts>0):\", np.sum(nCountsPerGene>0),)\n",
    "print(\"\\t nGenesPerCell: \", round(nGenesPerCell.min(),2), \" - \" , round(nGenesPerCell.max(),2))\n",
    "print(\"\\t Maximum value in matrix: \", np.nanmax(np.asmatrix(exprMat)))\n",
    "print(\"\\t Number of counts (in the dataset units) per gene:\", nCountsPerGene.min(), \" - \" ,nCountsPerGene.max())\n",
    "print(\"\\t Number of cells in which each gene is detected:\", nCellsPerGene.min(), \" - \" ,nCellsPerGene.max())\n",
    "\n",
    "## Set thresholds\n",
    "minCountsPerGene=3*.003*nCells # 3 counts in .3% of cells\n",
    "minSamples=.003*nCells # .3% of cells\n",
    "\n",
    "print(\"Used for filtering: \")\n",
    "print(\"\\t minCountsPerGene: \", round(minCountsPerGene,2))\n",
    "print(\"\\t minSamples: \", round(minSamples,2), \" (nCells:\",nCells,\")\")\n",
    "\n",
    "\n",
    "print(\"Genes that pass the filter:\")\n",
    "# First filter\n",
    "genesLeft_minReads=nCountsPerGene[nCountsPerGene>minCountsPerGene].index.tolist()\n",
    "print(\"\\t \", len(genesLeft_minReads), \"\\tgenes with counts per gene > \", round(minCountsPerGene,2))\n",
    "  \n",
    "# Second filter\n",
    "nCellsPerGene2=nCellsPerGene[genesLeft_minReads]\n",
    "genesLeft_minCells=nCellsPerGene2[nCellsPerGene2>minSamples].index.tolist()\n",
    "print(\"\\t \", len(genesLeft_minCells), \"\\tgenes detected in more than \", round(minSamples,2),\" cells\")\n",
    "\n",
    "# In db\n",
    "ctxDb_geneList=feather.read_feather(DATABASES_GLOB).columns[1:]\n",
    "genesKept=list(set(genesLeft_minCells).intersection(ctxDb_geneList))\n",
    "genesKept.sort()\n",
    "print(\"\\t \", len(genesKept), \"\\tgenes are in the motif database (the database contains info for\",len(ctxDb_geneList),\"genes)\")\n",
    "\n",
    "# Subst matrix\n",
    "exprMat_filtered=exprMat.loc[genesKept]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1772071017_F07</th>\n",
       "      <th>1772071017_E06</th>\n",
       "      <th>1772067066_C10</th>\n",
       "      <th>1772071017_B05</th>\n",
       "      <th>1772071014_E06</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1700030J22Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700030K09Rik</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700034H15Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700037C18Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700037H04Rik</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700047M11Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700048O20Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700049G17Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700052K11Rik</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700052N19Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               1772071017_F07  1772071017_E06  1772067066_C10  1772071017_B05  \\\n",
       "1700030J22Rik               0               1               0               0   \n",
       "1700030K09Rik               1               1               0               0   \n",
       "1700034H15Rik               0               0               0               0   \n",
       "1700037C18Rik               0               0               0               0   \n",
       "1700037H04Rik               2               4               3               4   \n",
       "1700047M11Rik               0               0               0               0   \n",
       "1700048O20Rik               0               0               0               0   \n",
       "1700049G17Rik               0               0               0               0   \n",
       "1700052K11Rik               3               0               0               0   \n",
       "1700052N19Rik               0               0               2               1   \n",
       "\n",
       "               1772071014_E06  \n",
       "1700030J22Rik               0  \n",
       "1700030K09Rik               0  \n",
       "1700034H15Rik               0  \n",
       "1700037C18Rik               0  \n",
       "1700037H04Rik               5  \n",
       "1700047M11Rik               0  \n",
       "1700048O20Rik               3  \n",
       "1700049G17Rik               0  \n",
       "1700052K11Rik               0  \n",
       "1700052N19Rik               1  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exprMat.iloc[100:110,25:30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13800, 3005)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exprMat_filtered.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save \n",
    "We suggest/recommend saving the filtered matrix as [.loom](https://linnarssonlab.org/loompy/) file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import loompy\n",
    "row_metadata={'Gene':list(exprMat_filtered.index)}\n",
    "col_metadata={'CellID':list(exprMat_filtered.columns.values)}\n",
    "loompy.create(exprMat_filtered_path, \n",
    "              exprMat_filtered.values,\n",
    "              row_attrs=row_metadata, \n",
    "              col_attrs=col_metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To reload/confirm that is saved propperly..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1772071017_F07</th>\n",
       "      <th>1772071017_E06</th>\n",
       "      <th>1772067066_C10</th>\n",
       "      <th>1772071017_B05</th>\n",
       "      <th>1772071014_E06</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1700030J22Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700030K09Rik</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700034H15Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700037C18Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700037H04Rik</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700047M11Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700048O20Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700049G17Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700052K11Rik</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700052N19Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               1772071017_F07  1772071017_E06  1772067066_C10  1772071017_B05  \\\n",
       "1700030J22Rik               0               1               0               0   \n",
       "1700030K09Rik               1               1               0               0   \n",
       "1700034H15Rik               0               0               0               0   \n",
       "1700037C18Rik               0               0               0               0   \n",
       "1700037H04Rik               2               4               3               4   \n",
       "1700047M11Rik               0               0               0               0   \n",
       "1700048O20Rik               0               0               0               0   \n",
       "1700049G17Rik               0               0               0               0   \n",
       "1700052K11Rik               3               0               0               0   \n",
       "1700052N19Rik               0               0               2               1   \n",
       "\n",
       "               1772071014_E06  \n",
       "1700030J22Rik               0  \n",
       "1700030K09Rik               0  \n",
       "1700034H15Rik               0  \n",
       "1700037C18Rik               0  \n",
       "1700037H04Rik               5  \n",
       "1700047M11Rik               0  \n",
       "1700048O20Rik               3  \n",
       "1700049G17Rik               0  \n",
       "1700052K11Rik               0  \n",
       "1700052N19Rik               1  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with loompy.connect(exprMat_filtered_path) as ds:\n",
    "    # ds.ra.keys(); ds.ca.keys()\n",
    "    exprMat=pd.DataFrame(ds[:,:], index=list(ds.ra['Gene']))\n",
    "    exprMat.columns=list(ds.ca['CellID'])\n",
    "\n",
    "exprMat.iloc[100:110,25:30]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Next:\n",
    "Now you are ready to un SCENIC... \n",
    "\n",
    "Template notebook:`pySCENIC_template.ipynb`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
